jsonrpc/websocket.go

```diff
@@ -95,52 +105,53 @@ func DefaultWebsocketConnParams() *WebsocketConnParams {
 
 type websocketConn struct {
 	conn     *websocket.Conn
-	rpc      *Server
+	buf      []byte
+	ctx      context.Context
 	params   *WebsocketConnParams
 	listener NewRequestListener
 }
 
-func newWebsocketConn(conn *websocket.Conn, rpc *Server, params *WebsocketConnParams, listener NewRequestListener) *websocketConn {
+func newWebsocketConn(ctx context.Context, conn *websocket.Conn, params *WebsocketConnParams, listener NewRequestListener) *websocketConn {
 	conn.SetReadLimit(params.ReadLimit)
 	return &websocketConn{
 		conn:     conn,
-		rpc:      rpc,
+		ctx:      ctx,
 		params:   params,
 		listener: listener,
 	}
 }
 
-func (wsc *websocketConn) ReadWriteLoop(ctx context.Context) error {
-	for {
-		// Read next message from the client.
-		_, r, err := wsc.conn.Read(ctx)
-		if err != nil {
-			return err
-		}
-
-		// TODO write responses concurrently. Unlike gorilla/websocket, nhooyr.io/websocket
-		// permits concurrent writes.
-
-		wsc.listener.OnNewRequest("any")
-		// Decode the message, call the handler, encode the response.
-		resp, err := wsc.rpc.Handle(ctx, r)
+func (wsc *websocketConn) Read(p []byte) (int, error) {
+	if wsc.buf == nil {
+		var err error
+		_, wsc.buf, err = wsc.conn.Read(wsc.ctx)
 		if err != nil {
-			// RPC handling issues should not affect the connection.
-			// Ignore the request and let the client close the connection.
-			// TODO: is there a better way to do this?
-			continue
-		}
-
-		// Send the message to the client.
-		if err := wsc.Write(ctx, resp); err != nil {
-			return err
+			return len(wsc.buf), err
 		}
 	}
+	var n int
+	if len(wsc.buf) > len(p) {
+		n = copy(p, wsc.buf[:len(p)])
+		wsc.buf = wsc.buf[len(p):]
+	} else {
+		n = copy(p, wsc.buf)
+		wsc.buf = nil
+	}
+	wsc.listener.OnNewRequest("any")
```

Comment:
I am not sure if this is the right place for this, moving it to after `HandleReader()` in the loop above makes more sense to me.